libc/dns/prototxt.c

```diff
@@ -0,0 +1,190 @@
+ * @param bufsize is the size of buf
+ * @param filepath is the location of the protocols file
+ *          (if NULL, uses /etc/protocols)
+ * @returns 0 on success, -1 on error
+ *
+ * @note aliases are not read from the file.
+ */
+int LookupProtoByNumber(const int protonum, char *buf, size_t bufsize,
+                        const char *filepath) {
+  FILE *f;
+  char *line;
+  char pathbuf[PATH_MAX];
+  const char *path;
+  size_t linesize;
+  int found;
+  char *name, *number, *comment, *tok;
+
+  if (!(path = filepath)) {
+    path = "/etc/protocols";
+    if (IsWindows()) {
+      path =
+          firstnonnull(GetNtProtocolsTxtPath(pathbuf, ARRAYLEN(pathbuf)), path);
+    }
+  }
+
+  if (bufsize == 0 || !(f = fopen(path, "r"))) {
+    return -1;
+  }
+  line = NULL;
+  linesize = 0;
+  found = 0;
+
+  while (found == 0 && (getline(&line, &linesize, f)) != -1) {
+    if ((comment = strchr(line, '#'))) *comment = '\0';
+    name = strtok_r(line, " \t\r\n\v", &tok);
+    number = strtok_r(NULL, " \t\r\n\v", &tok);
+    if (name && number && protonum == atoi(number)) {
+      strncpy(buf, name, bufsize);
+      found = 1;
+    }
+  }
+  free(line);
+
+  if (ferror(f)) {
+    errno = ferror(f);
+    return -1;
+  }
+  fclose(f);
+
+  if (!found) return -1;
+
+  return 0;
+}
+
+/**
+ * Opens and searches /etc/protocols to find number for a given name.
+ *
+ * @param protoname is a NULL-terminated string
+ * @param buf is a buffer to store the official name of the protocol
+ * @param bufsize is the size of buf
+ * @param filepath is the location of protocols file
+ *          (if NULL, uses /etc/protocols)
+ * @returns -1 on error, or
+ *          positive protocol number
+ *
+ * @note aliases are read from file for comparison, but not returned.
+ * @see LookupProtoByNumber
+ */
+int LookupProtoByName(const char *protoname, char *buf, size_t bufsize,
+                      const char *filepath) {
+  FILE *f;
+  char *line;
+  char pathbuf[PATH_MAX];
+  const char *path;
+  size_t linesize;
+  int found, result;
+  char *name, *number, *alias, *comment, *tok;
+
+  if (!(path = filepath)) {
+    path = "/etc/protocols";
+    if (IsWindows()) {
+      path =
+          firstnonnull(GetNtProtocolsTxtPath(pathbuf, ARRAYLEN(pathbuf)), path);
+    }
+  }
+
+  if (bufsize == 0 || !(f = fopen(path, "r"))) {
+    return -1;
+  }
+  line = NULL;
+  linesize = 0;
+  found = 0;
+  result = -1;
+
+  while (found == 0 && (getline(&line, &linesize, f)) != -1) {
+    if ((comment = strchr(line, '#'))) *comment = '\0';
+    name = strtok_r(line, " \t\r\n\v", &tok);
+    number = strtok_r(NULL, "/ \t\r\n\v", &tok);
+    if (name && number) {
+      alias = name;
+      while (alias && strcasecmp(alias, protoname) != 0)
+        alias = strtok_r(NULL, " \t\r\n\v", &tok);
+
+      if (alias) /* alias matched with protoname */
+      {
+        result = atoi(number);
+        strncpy(buf, name, bufsize);
```

Comment:
Consider `strlcpy` since that actually does the thing `strncpy` should have done. `strncpy` doesn't guarantee a NUL terminator so it can't be used to copy a string to a fixed size buffer with silent truncation. It's also possible to use `memccpy` for this purpose. We have a SWAR implementation of `memccpy` since I like that function a lot. However it's got a more cumbersome API since it requires that you check the return value.

https://github.com/jart/cosmopolitan/blob/fe881982b5da8d34999e8fdb13bd62c382391032/libc/str/memccpy.c#L29-L92

Reply:
`memccpy` failure handling is nice: `memccpy` returns NULL on failure, `strlcpy` returns the length copied, which I have to check with something else, and `strncpy` simply returns the destination pointer.